packages/gatsby/cache-dir/dev-loader.js

```diff
@@ -161,42 +166,78 @@ class DevLoader extends BaseLoader {
           }
         })
       }
+      return true
+    }
+    return false
+  }
+
+  markAsStale = dirtyQueryId => {
+    if (dirtyQueryId === `/dev-404-page/` || dirtyQueryId === `/404.html`) {
+      // those pages are not on demand so skipping
+      return
+    }
 
-      ___emitter.emit(`pageQueryResult`, newPageData)
+    const normalizedId = normalizePagePath(dirtyQueryId)
+
+    // We can't just delete items in caches, because then
+    // using history.back() would show dev-404 page
+    // due to our special handling of it in root.js (loader.isPageNotFound check)
+    // so instead we mark it as stale and instruct loader's async methods
+    // to refetch resources if they are marked as stale
+
+    const cachedPageData = this.pageDataDb.get(normalizedId)
+    if (cachedPageData) {
+      // if we have page data in cache, mark it as stale
+      this.pageDataDb.set(normalizedId, {
+        ...cachedPageData,
+        stale: true,
+      })
+    }
+
+    const cachedPage = this.pageDb.get(normalizedId)
+    if (cachedPage) {
+      // if we have page data in cache, mark it as stale
+      this.pageDb.set(normalizedId, {
+        ...cachedPage,
+        payload: { ...cachedPage.payload, stale: true },
+      })
     }
   }
 
-  handleStalePageDataMessage(msg) {
-    msg.payload.stalePageDataPaths.forEach(dirtyQueryId => {
-      if (dirtyQueryId === `/dev-404-page/` || dirtyQueryId === `/404.html`) {
-        // those pages are not on demand so skipping
-        return
-      }
+  handlePageQueryResultHotUpdate(msg) {
+    const updated = this.updatePageData(msg.payload.id, msg.payload.result)
+    if (updated) {
+      ___emitter.emit(`pageQueryResult`, msg.payload.result)
+    }
+  }
 
-      const normalizedId = normalizePagePath(dirtyQueryId)
+  handleStalePageDataMessage(msg) {
+    for (const dirtyQueryId of msg.payload.stalePageDataPaths) {
+      this.markAsStale(dirtyQueryId)
+    }
+  }
 
-      // We can't just delete items in caches, because then
-      // using history.back() would show dev-404 page
-      // due to our special handling of it in root.js (loader.isPageNotFound check)
-      // so instead we mark it as stale and instruct loader's async methods
-      // to refetch resources if they are marked as stale
+  handleStaleServerDataMessage() {
+    const activePath = normalizePagePath(location.pathname)
 
-      const cachedPageData = this.pageDataDb.get(normalizedId)
-      if (cachedPageData) {
-        // if we have page data in cache, mark it as stale
-        this.pageDataDb.set(normalizedId, {
-          ...cachedPageData,
-          stale: true,
-        })
+    // For now just invalidate every single page with serverData
+    for (const [key, value] of this.pageDataDb) {
+      if (value?.payload?.result?.serverData) {
+        this.markAsStale(key)
+      }
+      if (activePath === normalizePagePath(key)) {
```

Comment:
Note to self - check if this condition will work with query params (once other PR is merged)